File: src/net/merayen/elastic/ui/objects/node/Resizable.java
package net.merayen.elastic.ui.objects.node;

import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.event.UIEvent;
import net.merayen.elastic.ui.event.MouseEvent.Button;
import net.merayen.elastic.ui.util.MouseHandler;
import net.merayen.elastic.util.Point;

public class Resizable extends UIObject {
	public interface Handler {
		public void onResize();
	}

	private final Handler handler;
	private final UINode node;
	private MouseHandler mouse_handler;

	private float start_width, start_height;

	public Resizable(UINode node, Handler handler) {
		this.node = node;
		this.handler = handler;
	}

	@Override
	protected void onInit() {
		mouse_handler = new MouseHandler(this, Button.LEFT);
		mouse_handler.setHandler(new MouseHandler.Handler() {
			@Override
			public void onMouseDown(Point position) {
				start_width = node.width;
				start_height = node.height;
			}

			@Override
			public void onMouseDrag(Point position, Point offset) {
				node.width = start_width + offset.x;
				node.height = start_height + offset.y;
				handler.onResize();
			}
		});
	}

	@Override
	protected void onDraw() {
		super.onDraw();

		draw.setColor(200, 200, 200);
		draw.setStroke(0.5f);
		for(int i = 1; i < 8; i+=2)
			draw.line(node.width - 2 * i, node.height, node.width, node.height - 2 * i);
	}

	@Override
	protected void onEvent(UIEvent e) {
		mouse_handler.handle(e);
	}
}

File: src/net/merayen/elastic/ui/objects/node/Resizable.kt
package net.merayen.elastic.ui.objects.node

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.event.MouseEvent.Button
import net.merayen.elastic.ui.util.MouseHandler
import net.merayen.elastic.util.Point

class Resizable(private val node: UINode, private val handler: Handler) : UIObject() {
	private var mouseHandler: MouseHandler? = null

	private var startWidth: Float = 0.toFloat()
	private var startHeight: Float = 0.toFloat()

	interface Handler {
		fun onResize()
	}

	override fun onInit() {
		mouseHandler = MouseHandler(this, Button.LEFT)
		mouseHandler!!.setHandler(object : MouseHandler.Handler() {
			override fun onMouseDown(position: Point) {
				startWidth = node.getWidth()
				startHeight = node.getHeight()
			}

			override fun onMouseDrag(position: Point, offset: Point) {
				node.layoutWidth = startWidth + offset.x
				node.layoutHeight = startHeight + offset.y
				handler.onResize()
			}
		})
	}

	override fun onDraw(draw: Draw) {
		super.onDraw(draw)

		draw.setColor(200, 200, 200)
		draw.setStroke(0.5f)
		var i = 1
		while (i < 8) {
			draw.line(node.getWidth() - 2 * i, node.getHeight(), node.getWidth(), node.getHeight() - 2 * i)
			i += 2
		}
	}

	override fun onEvent(e: UIEvent) {
		mouseHandler!!.handle(e)
	}
}
